pkg/algorithm/hived_algorithm.go

```diff
@@ -472,137 +474,156 @@ func (h *HivedAlgorithm) setHealthyNode(nodeName string) {
+		klog.Infof("Cell type %v (chain %v level %v) now has sufficient healthy cells (%v) for "+
+			"allocating the total free cells of all the VCs (%v). No VC cell will be doomed to be bad.",
+			h.cellTypes[chain][level], chain, level,
+			h.totalLeftCellNum[chain][level]-int32(len(h.badFreeCells[chain][level])),
+			h.allVCFreeCellNum[chain][level]+h.doomedBadCellNum[chain][level])
+	}
+	h.tryUnbindDoomedBadCell(chain, level)
 }
 
-// updateVCDoomedBadCells checks if the healthy cells in the physical cluster can satisfy the free cells
+// tryBindDoomedBadCell checks if the healthy cells in the physical cluster can satisfy the free cells
 // in each VC. If not, we will mark some of the virtual cells in the VC as bad (i.e., doomed to be bad).
-func (h *HivedAlgorithm) updateVCDoomedBadCells(c CellChain, l CellLevel) {
-	for vcName, vcFreeCellNum := range h.vcFreeCellNum {
-		if _, ok := vcFreeCellNum[c][l]; !ok {
+func (h *HivedAlgorithm) tryBindDoomedBadCell(c CellChain, l CellLevel) {
+	for vcName, vcFreeNum := range h.vcFreeCellNum {
+		if _, ok := vcFreeNum[c]; !ok {
 			continue
 		}
-		prevDoomedBadNum := h.vcDoomedBadCellNum[vcName][c][l]
-		h.vcDoomedBadCellNum[vcName][c][l] = vcFreeCellNum[c][l] - h.totalLeftCellNum[c][l] + h.badFreeCellNum[c][l]
-		if h.vcDoomedBadCellNum[vcName][c][l] <= 0 {
-			h.vcDoomedBadCellNum[vcName][c][l] = 0
-			if prevDoomedBadNum > 0 {
-				klog.Infof("Cell type %v (chain %v level %v) now has sufficient healthy cells (healthy %v, bad %v) "+
-					"for allocating the free cells of the VC %v (%v).",
-					h.cellTypes[c][l], c, l,
-					h.totalLeftCellNum[c][l]-h.badFreeCellNum[c][l], h.badFreeCellNum[c][l],
-					vcName, vcFreeCellNum[c][l])
-			}
-		} else {
-			klog.Warningf("Cell type %v (chain %v level %v) now has fewer healthy cells (healthy %v, bad %v) "+
-				"than the free cells of the VC %v (%v). Certain cells in the VC are doomed to be bad.",
+		for vcFreeNum[c][l] >
+			h.totalLeftCellNum[c][l]-int32(len(h.badFreeCells[c][l])) {
+			pc := h.badFreeCells[c][l][0].(*PhysicalCell)
+			vc := getUnboundVirtualCell(h.vcSchedulers[vcName].getNonPinnedPreassignedCells()[c][l])
+			pc.SetVirtualCell(vc)
+			vc.SetPhysicalCell(pc)
+			klog.Infof("Virtual cell %v is bound to physical cell %v", vc.GetAddress(), pc.GetAddress())
+			klog.Warningf("Cell type %v (chain %v level %v) now has fewer healthy cells (%v) "+
+				"than the free cells of the VC %v (%v). "+
+				"Cell %v is doomed to be bad and bound to %v.",
 				h.cellTypes[c][l], c, l,
-				h.totalLeftCellNum[c][l]-h.badFreeCellNum[c][l], h.badFreeCellNum[c][l],
-				vcName, vcFreeCellNum[c][l])
+				h.totalLeftCellNum[c][l]-int32(len(h.badFreeCells[c][l])),
+				vcName, vcFreeNum[c][l]+int32(len(h.vcDoomedBadCells[vcName][c][l])),
+				vc.GetAddress(), pc.GetAddress())
+			h.vcDoomedBadCells[vcName][c][l] = append(h.vcDoomedBadCells[vcName][c][l], pc)
+			h.doomedBadCellNum[c][l]++
+			h.allocatePreassignedCell(pc, vcName)
 		}
```

Comment:
Can we reuse more funcs like the way we bind used cells?

Reply:
The other code in allocateGpu is mostly for setting priorities, updating used gpu number
No need to reuse them for bad cells